Automattic-Tracks-iOS/Crash Logging/CrashLogging.swift

```diff
@@ -2,178 +2,251 @@ import Foundation
+        event.extra?["localized-description"] = error.localizedDescription
+
+        event.timestamp = Date()
+
+        SentrySDK.capture(event: event)
+        dataProvider.didLogErrorCallback?(event)
+    }
+
+    /// Writes a message to the Crash Logging system, and includes a stack trace.
+    ///
+    /// - Parameters:
+    ///   - message: The message
+    ///   - properties: A dictionary containing additional information about this error
+    ///   - level: The level of severity to report in Sentry (`.info` by default)
+    func logMessage(_ message: String, properties: [String: Any]? = nil, level: SentryLevel = .info) {
+
+        let event = Event(level: level)
+        event.message = SentryMessage(formatted: message)
+        event.extra = properties
         event.timestamp = Date()
 
-        Client.shared?.snapshotStacktrace {
-            Client.shared?.appendStacktrace(to: event)
+        SentrySDK.capture(event: event)
+        dataProvider.didLogMessageCallback?(event)
+    }
+
+    /// Sends an `Event` to Sentry and triggers a callback on completion
+    func logErrorImmediately(_ error: Error, userInfo: [String: Any]? = nil, level: SentryLevel = .error, callback: @escaping (Result<Bool, Error>) -> Void) throws {
+        try logErrorsImmediately([error], userInfo: userInfo, level: level, callback: callback)
+    }
+
+    func logErrorsImmediately(_ errors: [Error], userInfo: [String: Any]? = nil, level: SentryLevel = .error, callback: @escaping (Result<Bool, Error>) -> Void) throws {
+
+        var serializer = SentryEventSerializer(dsn: dataProvider.sentryDSN)
+
+        errors.forEach {
+            let event = Event(level: level)
+            event.message = SentryMessage(formatted: $0.localizedDescription)
+            event.timestamp = Date()
+            event.extra = userInfo ?? ($0 as NSError).userInfo
+            event.user = dataProvider.currentUser?.sentryUser
+
+            serializer.add(event: tryAddingStackTrace(to: event))
         }
 
-        Client.shared?.send(event: event)
-        sharedInstance.dataProvider?.didLogErrorCallback?(event)
+        guard let requestBody = try? serializer.serialize() else {
+            DDLogError("⛔️ Unable to send errors to Sentry – error could not be serialized. Attempting to schedule delivery for another time.")
+            errors.forEach {
+                SentrySDK.capture(error: $0)
+            }
+            return
+        }
+
+        let dsn = try SentryDsn(string: dataProvider.sentryDSN)
+        guard let authString = dsn.getAuthString() else {
+            throw Errors.unableToConstructAuthStringError
+        }
+
+        var request = URLRequest(url: dsn.getEnvelopeEndpoint())
+        request.httpMethod = "POST"
+        request.httpBody = requestBody
+        request.addValue(authString, forHTTPHeaderField: "X-Sentry-Auth")
+
+        URLSession.shared.dataTask(with: request) { (responseData, urlResponse, error) in
```

Comment:
Future refinement suggestion: what do you think of injecting a `URLSession` instance at `init` time (maybe with default `.shared` value) to avoid this direct singleton access?

It could eventually help us to test the callback calling behavior.

Reply:
Yep, agreed. Eventually I'd like for all of the `Tracks` functionality to be able to use an injected session that can be defined by the hosting app – would make sense to roll out to all the things for sure.